sdk/dotnet/Pulumi/Resources/InputArgs.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright 2016-2019, Pulumi Corporation
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using Pulumi.Serialization;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Base type for all input argument classes.
+    /// </summary>
+    public abstract class InputArgs
+    {
+        private readonly ImmutableArray<InputInfo> _inputInfos;
+
+        private protected abstract void ValidateMember(Type memberType, string fullName);
+
+        protected InputArgs()
+        {
+            var fieldQuery =
+                from field in this.GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)
+                let attr = field.GetCustomAttribute<InputAttribute>()
+                where attr != null
+                select (attr, memberName: field.Name, memberType: field.FieldType, getValue: (Func<object, object?>)field.GetValue);
+
+            var propQuery =
+                from prop in this.GetType().GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)
+                let attr = prop.GetCustomAttribute<InputAttribute>()
+                where attr != null
+                select (attr, memberName: prop.Name, memberType: prop.PropertyType, getValue: (Func<object, object?>)prop.GetValue);
+
+            var all = fieldQuery.Concat(propQuery).ToList();
+
+            foreach (var (attr, memberName, memberType, getValue) in all)
+            {
+                var fullName = $"[Input] {this.GetType().FullName}.{memberName}";
+                ValidateMember(memberType, fullName);
```

Comment:
the only real change is that we don't require all fields/props to be `Input<T>` at the base level.